lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -214,6 +214,43 @@ TypeChecker::lookupUnqualifiedType(DeclContext *dc, DeclName name,
                                    NameLookupOptions options) {
   SmallVector<TypeDecl *, 1> decls;
 
+  // CodingKeys is a special type. Although not a protocol requirement of the
+  // Encodable/Decodable protocols, it is synthesized as part of
+  // Encodable/Decodable conformance. Unqualified lookup may occur before the
+  // type is synthesized, though, in which case, we need to make sure it gets
+  // synthesized.
+  //
+  // If this is what we're looking for, then we need to attempt to cause the
+  // type to be synthesized before performing the lookup.
+  if (name.isSimpleName(Context.Id_CodingKeys)) {
+    // UnqualifiedLookup below has no knowledge of type checking and has no way
+    // to expand protocol conformance, which is what we'd normally need.
+    //
+    // Instead, we need to iterate up the AST first here, attempting to perform
+    // qualified lookups first.
+    auto *currentDC = dc;
+    while (currentDC) {
+      auto *nominal =
+        currentDC->getAsNominalTypeOrNominalTypeExtensionContext();
+      if (nominal) {
+        // We've found a nominal context to look in. Let's preform a qualified
+        // lookup here (which will cause CodingKeys synthesis if possible).
+        auto results = lookupMemberType(nominal->getDeclContext(),
+                                        nominal->getDeclaredInterfaceType(),
+                                        Context.Id_CodingKeys, options);
+        if (results.size() > 0) {
+          for (auto &result : results) {
+            decls.push_back(result.first);
+          }
+
+          return decls;
```

Comment:
Yeah, what I really wanted to do is push these things into `lib/AST/NameLookup.cpp` so we get finer-grained control, but again, there's no knowledge in there of type checking and protocol conformance... 😕  But yes, you're right, we can't blindly do lookup like this. Local (unqualified) types should win.